python/tk_desktop/desktop_engine_project_implementation.py

```diff
@@ -72,67 +113,111 @@ def post_app_init(self):
-            try:
-                self.proxy.call_no_response("destroy_app_proxy")
-            except EOFError:
-                # it is ok if it is already shut down on the other side
-                pass
-
-            self.proxy.close()
-
-        # close down our server thread
-        if self.msg_server is not None:
-            self.msg_server.close()
+        self._project_comm.shut_down()
 
-    def trigger_callback(self, namespace, command, *args, **kwargs):
+    def _trigger_callback(self, namespace, command, *args, **kwargs):
         callback = self.__callback_map.get((namespace, command))
         try:
             callback(*args, **kwargs)
         except Exception:
             self._engine.log_error("Error calling %s::%s(%s, %s):\n%s" % (
                 namespace, command, args, kwargs, traceback.format_exc()))
 
-    def signal_disconnect(self):
-        if self._engine.has_ui:
+    def _signal_disconnect(self):
+        # If the user is quick enough and steps out of the project right after launching it, the engine
+        # will be initialized but the app won't be.
+        with self._lock:
+            # QtGui is accessible in non-qt-based environments through the QtProxy, but won't be accessed if has_ui is
+            # False.
             from tank.platform.qt import QtGui
+            # If the engine reports having qt, but the QApplication hasn't been initialized yet, we want to make sure
+            # that the tear down is going to be coherent and
+            self._engine.has_ui = True if (self._engine.has_ui and QtGui.QApplication.instance()) else False
 
+        if self._engine.has_ui:
             app = QtGui.QApplication.instance()
-            self.connected = False
 
             top_level_windows = app.topLevelWidgets()
-            if top_level_windows:
-                self._engine.log_debug("Disconnected with open windows, setting quit on last window closed.")
+            sgtk.platform.current_engine()
+            # Bugs in client code (or Toolkit) can cause a leak of a top-level dialog that is already
+            # closed. So if any top level dialog is visible, wait for it to close.
+            opened_windows = filter(lambda w: w.isVisible(), top_level_windows)
+            if opened_windows:
+                self._engine.log_debug("The following top level widgets are still visible:")
+                for w in opened_windows:
+                    self._engine.log_debug(str(w))
+                self._engine.log_debug("Process will quit only when the last window is closed.")
                 app.setQuitOnLastWindowClosed(True)
             else:
                 self._engine.log_debug("Quitting on disconnect")
                 app.quit()
         else:
-            # just close down the message thread otherwise
-            self._engine.log_debug("Disconnect with no UI.  Shutting down")
-            self.msg_server.close()
+            self._project_comm.shut_down()
+
+    def start_app(self):
```

Comment:
This was refactored from the bootstrap_utilities so init and teardown live side by side.